src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -206,6 +222,33 @@ private static uint HandleEventNewStream(State state, ref ConnectionEvent connec
 
         private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent connectionEvent)
         {
+            if (connectionEvent.Data.StreamsAvailable.UniDirectionalCount > 0)
+            {
+                TaskCompletionSource? tcs;
+                lock (state)
+                {
+                    tcs = state.NewUnidirectionalStreamsAvailable;
+                    state.NewUnidirectionalStreamsAvailable = null;
+                }
+                if (tcs is not null)
+                {
+                    tcs.SetResult();
+                }
+            }
+            if (connectionEvent.Data.StreamsAvailable.BiDirectionalCount > 0)
+            {
+                TaskCompletionSource? tcs;
+                lock (state)
+                {
+                    tcs = state.NewBidirectionalStreamsAvailable;
+                    state.NewBidirectionalStreamsAvailable = null;
+                }
+                if (tcs is not null)
+                {
+                    tcs.SetResult();
+                }
+            }
```

Comment:
would it be beneficial to do both uni and bidi updates inside a single lock, not in two subsequent?
```suggestion
            TaskCompletionSource? unidirectionalTcs;
            TaskCompletionSource? bidirectionalTcs;
            lock (state)
            {
                if (connectionEvent.Data.StreamsAvailable.UniDirectionalCount > 0)
                {
                    unidirectionalTcs = state.NewUnidirectionalStreamsAvailable;
                    state.NewUnidirectionalStreamsAvailable = null;
                }
                if (connectionEvent.Data.StreamsAvailable.BiDirectionalCount > 0)
                {
                    bidirectionalTcs = state.NewBidirectionalStreamsAvailable;
                    state.NewBidirectionalStreamsAvailable = null;
                }
            }
            if (unidirectionalTcs is not null)
            {
                unidirectionalTcs.SetResult();
            }
            if (bidirectionalTcs is not null)
            {
                bidirectionalTcs.SetResult();
            }
```

Reply:
Anyone else has an opinion on this? Will single lock be more efficient here?

Reply:
I decided to go with the suggestion and not wait with it.